HW-05 DataBinding 적용

wkdehdlr opened the pull request:
늦어서 죄송합니다..!

shareitDOIK committed:
HW05 extension gradle 삭제

shareitDOIK committed:
extension import 제거

shareitDOIK committed:
xml 파일 layout 추가

shareitDOIK committed:
databinding 적용

zozero94 commented on wkdehdlr/CoinHelper/app/src/main/java/kr/schoolsharing/coinhelper/tasks/UpbitRVAdapter.kt:
```diff
@@ -4,18 +4,21 @@ import android.graphics.Color
+        binding = MainRvItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+        return Holder(binding)
+    }
     override fun onBindViewHolder(holder: Holder, position: Int) = holder.bind(itemList[position])
```
바인드하는 부분에서 executePendingBindings를 호출하는걸 
[권장]이 아닌 [필수] 라고 했던것 같습니다
https://developer.android.com/topic/libraries/data-binding/generated-binding?hl=ko#advanced_binding

wkdehdlr replied:
[54cf094](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/54cf0948f7a2078b927f14cbb439a29debd9e7db) 수정했습니다!

wkdehdlr replied:
요즘 Rx를 공부하는중인데 재영님 코드에 질문이 있습니다!
`

      repository.getMarkets()
            .map { list ->
                list.filter { filterData
                    ->
                    filterData.market.startsWith(index.name)
                }
                    .joinToString(",") { separateData ->
                        separateData.market
                    }
            }
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                {
                    marketName = it
                    tickerRequest()
                },
                {
                    view.showError(it.message)
                }).also { dispose = it }
    }`
CoinPresenter에서 observeOn전에 연산자로 데이터 가공하는 부분은 
RemoteDataSourceImpl에서 

` 
 api.getMarkets()

            .subscribeOn(Schedulers.io())`
로 되어있는데 여기서 설정한 (subscribeOn으로 설정한) 스케쥴러에서 연산자 작업도 이루어지는건가요??

zozero94 replied:
subscribeOn 설정 이후에 작동되는 작업이 io thread에서 작동하는거고
observeOn 설정 이후에 작동되는 작업이 Android main thread에서 작동합니다.

remoteDataSource에서 subscribeOn 스케쥴러를 지정해준다음 리턴값으로 스트림을 넘겨주면
그걸 Repository에서 local을 사용할지 remote를 사용할지 정해서 Presenter로 넘겨주는 방식입니다. 

그러면 remoteDataSource에서 정의된 스케쥴러(subscribeOn) 을 통해 작업이 이루어지고
마지막 presenter에서 구독하기 전에 메인스레드로 변경하여 view에 뿌려줄 수 있는 형태로 가공하는거죠 ㅇ_ㅇ...

wkdehdlr replied:
답변 감사합니다!

zozero94 reviewed (commented):
이번꺼는 kotlin extension만 제외하고 데이터바인딩만 엮는거라 딱히 리뷰할게 없네요!

shareitDOIK committed:
executePedingBindings() 추가

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/java/kr/schoolsharing/coinhelper/tasks/MainActivity.kt:
```diff
@@ -2,23 +2,25 @@ package kr.schoolsharing.coinhelper.tasks
     }
 
     private fun adaptPager() {
         PagerAdapter(supportFragmentManager).apply {
```
apply 보다는 with 나 run 이 적합해보입니다.

wkdehdlr replied:
객체를 생성함과 동시에 연속된 작업을 처리할 때 apply를 사용한다고 알고있어서 apply를 사용했는데 제가 잘못알고있나요????......

BeokBeok replied:
apply 는 수신객체를 반환합니다.
하지만, 본 코드에는 수신객체를 사용하지 않아 apply 보다는 with나 run이 적합해보여 말씀드렸습니다.

wkdehdlr replied:
[7d7bb2a](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/7d7bb2a6bea033fd8537f2f5bbbc1c41a1edb426) 수정했습니다. 이번을 계기로 정확히 이해했습니다. 감사합니다!

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/java/kr/schoolsharing/coinhelper/tasks/UpbitFragment.kt:
```diff
@@ -25,9 +28,15 @@ class UpbitFragment : Fragment(), UpbitContract.View {
+            false
+        )
+        return binding.root
+//        return inflater.inflate(R.layout.fragment_coinrecycler, container, false)
```
의미없는 주석은 제거해주세요.
또한, onViewCreated 함수는 사용하지 않는 것으로 보입니다.

wkdehdlr replied:
[f0f792d](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/f0f792d2a2e61a3eb8615ee645d9c64fe30d7eae) 수정했습니다

BeokBeok replied:
주석 처리하지 마시고 제거해주시고, 코맨트가 필요하다면 커밋 로그로 남겨주세요.

wkdehdlr replied:
[c1577fc](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/c1577fca904d48d49c2c830fec310d58fd42a6a9) 수정했습니다. 코드상에 아예 주석을 안남기는게 맞는 방향인건가요?

BeokBeok replied:
주석을 남기지 않고 코드만 보고 이해할 수 있는 것이 가장 좋고
로직이 복잡한 경우에는 이해를 돕기 위한 일부 주석을 넣는 것은 좋습니다.

하지만, 그 것이 아니라면 제거하는 것이 좋습니다.
왜냐하면 주석 처리를 해 놓고 신경쓰지 않으면, 
해당 주석은 영원히 남아있을 가능성이 높기 때문입니다.

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/java/kr/schoolsharing/coinhelper/tasks/UpbitFragment.kt:
```diff
@@ -39,13 +48,13 @@ class UpbitFragment : Fragment(), UpbitContract.View {
 
-        CoinRecyclerView.apply {
+
+        binding.CoinRecyclerView.apply {
```
apply 보다 with가 적합해보입니다.

wkdehdlr replied:
[7d7bb2a](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/7d7bb2a6bea033fd8537f2f5bbbc1c41a1edb426) 수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/java/kr/schoolsharing/coinhelper/tasks/UpbitRVAdapter.kt:
```diff
@@ -4,20 +4,26 @@ import android.graphics.Color
 
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder = Holder(parent)
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
+        binding = MainRvItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
```
가로 줄이 길기 때문에 함수 관련 convention 을 적용해주세요.
https://kotlinlang.org/docs/reference/coding-conventions.html#method-call-formatting

wkdehdlr replied:
[d36c104](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/d36c104fd16af1b45149e0090368fd89ede9cd27)  수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/java/kr/schoolsharing/coinhelper/tasks/UpbitRVAdapter.kt:
```diff
@@ -4,20 +4,26 @@ import android.graphics.Color
-    override fun onBindViewHolder(holder: Holder, position: Int) = holder.bind(itemList[position])
+    override fun onBindViewHolder(holder: Holder, position: Int) {
+        holder.bind(itemList[position])
+        binding.executePendingBindings()
```
`binding.executePendingBindings()`는 왜 사용하신건가요?

wkdehdlr replied:
재영님이 남겨주신 리뷰를 참고해서 반영했습니다.

BeokBeok replied:
재영님이 남겨주신 링크를 찾아가서 보았는데, 필수라고 명시된 부분을 찾지 못하였습니다.

> When a variable or observable object changes, the binding is scheduled to change before the next frame. **There are times, however, when binding must be executed immediately. To force execution, use the executePendingBindings() method.**

위의 글을 보아, 바인딩을 강제로 즉시 실행해야하는 경우에 사용하는 것으로 보이는데
놓친 부분이 있다면 피드백 부탁 드립니다.

wkdehdlr replied:
재영님 피드백 해주신 부분 통해서 확인했습니다~

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,65 +1,69 @@
-            android:layout_width="wrap_content"
+        <LinearLayout
+            android:id="@+id/title"
+            android:layout_width="match_parent"
```
Constraint Layout 자식 뷰의 match_parent는 0dp로 변경해주세요

wkdehdlr replied:
[e900a60](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/e900a603b90787651848068ea957b64c7af3519c) 수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,65 +1,69 @@
+                android:layout_height="wrap_content"
+                android:layout_weight="1"
+                android:gravity="center"
+                android:text="코인명" />
```
가급적이면 strings.xml 에 있는 리소스를 사용하여 참조해주세요.

wkdehdlr replied:
[4e75596](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/4e755969634b2c9a99ca666d2a795ec6667e7242) 수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,65 +1,69 @@
-            android:gravity="center"
-            android:text="현재가" />
+            <TextView
+                android:layout_width="wrap_content"
```
weight 속성을 사용하셨는데 wrap_content 를 사용하신 이유가 있나요?

wkdehdlr replied:
[7cd3e22](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/7cd3e22f9b17cbbd859773db9075bf1f5f39a00b) 놓치고 있었네요..! 수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,65 +1,69 @@
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@+id/title"></androidx.viewpager.widget.ViewPager>
```
ViewPager 의 내부 뷰가 없다면, `/>` 로 닫아주세요
또한, id를 생성하는 것이 아니라면 +를 제거해주세요.
```suggestion
            app:layout_constraintTop_toBottomOf="@id/title"/>
```

wkdehdlr replied:
[f000a05](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/f000a05d5f08c402f70fd8776182ce690a7c158e) 수정했습니다.  'id를 생성하는 것이 아니라면 +를 제거해주세요' 가 잘 이해가지 않습니다. 조금 더 설명을 부탁드려도 될까요..?

BeokBeok replied:
보통 뷰의 id를 지정할 때에는 다음과 같이 작성합니다.
`android:id="@+id/title`
즉, + 를 넣어줌으로써 id를 생성함과 동시에 설정을 해줍니다.

하지만 뷰의 id를 참조를 할 때에는 id를 생성하는것이 아니기 때문에
+를 제거하고 사용합니다.

wkdehdlr replied:
[93d92e9](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/93d92e95353c56abc303b9efa4fd1dcc4c2a2443) 수정했습니다. 이번에 새로 알게 된 사실입니다. 감사합니다!

BeokBeok replied:
id 속성에는 +를 넣어 주시고, 그 외의 속성에는 +를 제거해주세요.
`android:id="@+id/title`
`app:layout_constraintTop_toBottomOf="@id/title"`

wkdehdlr replied:
[9e518cf](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/9e518cf46561f899cf130df5dbced4378a115f3f) 수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/res/layout/fragment_coinrecycler.xml:
```diff
@@ -1,15 +1,19 @@
+    xmlns:tools="http://schemas.android.com/tools">
+
+    <FrameLayout
+        android:id="@+id/contentFrame"
```
안드로이드 리소스 관련해서는 snake case 를 권장합니다.
https://medium.com/mindorks/android-resource-naming-convention-42e4e8026614

wkdehdlr replied:
[63d68f5](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/63d68f542f7920fd0f59154c543a07e886a4da94) 수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/res/layout/fragment_coinrecycler.xml:
```diff
@@ -1,15 +1,19 @@
+        android:layout_height="match_parent"
+        tools:context=".tasks.UpbitFragment">
+
+        <!-- TODO: Update blank fragment layout -->
```
의미 없는 TODO는 제거해주세요

wkdehdlr replied:
[f0f792d](https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/198/commits/f0f792d2a2e61a3eb8615ee645d9c64fe30d7eae) 수정했습니다

BeokBeok commented on wkdehdlr/CoinHelper/app/src/main/res/layout/main_rv_item.xml:
```diff
@@ -10,31 +13,34 @@
+        <TextView
             android:id="@+id/current"
             android:gravity="center_vertical"
             android:textAlignment="textEnd"
```
해당 속성값에 오류가 있습니다.

BeokBeok reviewed (changes requested):
안녕하세요. 스터디 2기 강현석이라고 합니다.
부족하지만 잘 부탁드립니다. ^~^

shareitDOIK committed:
TODO 제거, 주석 정리, onViewCreated 주석처리

shareitDOIK committed:
Method call formatting 컨벤션 적용

shareitDOIK committed:
ViewPager empty body 처리

shareitDOIK committed:
string.xml 리소스 추가

shareitDOIK committed:
activity_main.xml  TextView layout_width 속성 wrap_content 수정

shareitDOIK committed:
ConstraintLayout match_parent->0dp 수정

shareitDOIK committed:
apply -> with 수정

shareitDOIK committed:
주석제거
       onActivityCreated 에서는 activity가 붙어있다고 확신할 수 있음 -> onViewCreted는             스쳐지나가는 정도임
           override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
                super.onViewCreated(view, savedInstanceState)
           }

shareitDOIK committed:
android:id수정//+제거

shareitDOIK committed:
안드로이드 리소스 snake case 적용

shareitDOIK committed:
@id 수정

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋

zozero94 reviewed (approved):
수고하셨습니다.

BeokBeok reviewed (approved):
LGTM:ok_hand:
